File: ios/Sources/KlinikLabFeature/LabTrendModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum LabTrendPhase: Sendable, Equatable {
    case loading
    case loaded
    /// Nothing confirmed yet. Not a failure: results exist only after review.
    case empty
    case notFound
    case failed(String)
}

public struct LabTrendState: Sendable, Equatable {
    public var phase: LabTrendPhase = .loading
    public var trends: [AnalyteTrend] = []
    /// Confirmed values that need looking at now, kept separate from the charts.
    public var critical: [LabResult] = []
    public var selected: String?
    public var error: String?

    public var selectedTrend: AnalyteTrend? {
        trends.first { $0.id == selected } ?? trends.first
    }

    public init() {}
}

/// The lab trend screen: one series per analyte, with its reference band.
public actor LabTrendModel {
    private let api: LabAPI
    private let subject: RecordSubject
    /// Asking the AI layer for a reading of these results. Nil on the
    /// patient's own copy, where there is nobody to sign one off.
    private let reports: ReportsAPI?

    private(set) public var state = LabTrendState()

    public init(api: LabAPI, subject: RecordSubject, reports: ReportsAPI? = nil) {
        self.api = api
        self.subject = subject
        self.reports = reports
    }

    public var canInterpret: Bool {
        guard reports != nil else { return false }

        if case .patient = subject { return true }

        return false
    }

    /**
     * Asks the AI layer to read these results (spec M5).
     *
     * The answer does not come back here. It goes into the review queue, where
     * a clinician signs it off before a patient sees anything — which is the
     * rule that makes asking for one safe at all, and a screen that displayed
     * the text immediately would route around it.
     */
    public func requestInterpretation() async -> Bool {
        guard let reports, case .patient(let id) = subject else { return false }

        state.error = nil

        do {
            _ = try await reports.interpretLabs(patientId: id)

            return true
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return false
    }

    public func currentState() -> LabTrendState { state }

    public func load() async {
        state.phase = .loading

        do {
            // Fetched together so a clinician's screen cannot show a chart
            // while still unaware of a critical value on the same patient.
            //
            // A patient looking at their own results does not get the critical
            // list. Spec M16 puts a doctor's review between an analyser and a
            // patient on purpose, and "CRITICAL" in red with nobody to explain
            // it is the outcome that review exists to prevent. The confirmed
            // trend is theirs to see; the alarm is the clinic's to act on.
            async let trends = api.trends(subject: subject)

            state.trends = try await trends

            if case .patient(let id) = subject {
                state.critical = try await api.critical(patientId: id)
            } else {
                state.critical = []
            }
            state.selected = state.trends.first?.id
            state.phase = state.trends.isEmpty ? .empty : .loaded
        } catch let error as APIError {
            if case .notFound = error {
                state.phase = .notFound
            } else {
                state.phase = .failed(L10n.message(for: error))
            }
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func select(_ id: String) {
        state.selected = id
    }
}

File: android/feature/lab/src/main/kotlin/xyz/klinik/feature/lab/LabTrendModel.kt
package xyz.klinik.feature.lab

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.RecordSubject
import xyz.klinik.network.AnalyteTrend
import xyz.klinik.network.ApiError
import xyz.klinik.network.LabApi
import xyz.klinik.network.LabResult
import xyz.klinik.network.messageKey

sealed interface LabTrendPhase {
    data object Loading : LabTrendPhase
    data object Loaded : LabTrendPhase

    /** Nothing confirmed yet. Not a failure: results exist only after review. */
    data object Empty : LabTrendPhase

    data object NotFound : LabTrendPhase
    data class Failed(val messageKey: String) : LabTrendPhase
}

data class LabTrendState(
    val phase: LabTrendPhase = LabTrendPhase.Loading,
    val trends: List<AnalyteTrend> = emptyList(),
    /** Confirmed values that need looking at now, kept out of the charts. */
    val critical: List<LabResult> = emptyList(),
    val selected: String? = null,
) {
    val selectedTrend: AnalyteTrend?
        get() = trends.firstOrNull { it.id == selected } ?: trends.firstOrNull()
}

/** The lab trend screen: one series per analyte, with its reference band. */
class LabTrendModel(
    private val api: LabApi,
    private val subject: RecordSubject,
) {
    private val _state = MutableStateFlow(LabTrendState())
    val state: StateFlow<LabTrendState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = LabTrendPhase.Loading)

        try {
            // Fetched together so a clinician's screen cannot show a chart
            // while still unaware of a critical value on the same patient.
            //
            // A patient looking at their own results does not get the critical
            // list. Spec M16 puts a doctor's review between an analyser and a
            // patient on purpose, and "CRITICAL" in red with nobody there to
            // explain it is the outcome that review exists to prevent.
            coroutineScope {
                val trends = async { api.trends(subject) }
                val critical = async {
                    if (subject is RecordSubject.Patient) api.critical(subject.id) else emptyList()
                }

                val series = trends.await()

                _state.value = _state.value.copy(
                    trends = series,
                    critical = critical.await(),
                    selected = series.firstOrNull()?.id,
                    phase = if (series.isEmpty()) LabTrendPhase.Empty else LabTrendPhase.Loaded,
                )
            }
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.NotFound) {
                    LabTrendPhase.NotFound
                } else {
                    LabTrendPhase.Failed((error as? ApiError)?.messageKey() ?: "error.server")
                },
            )
        }
    }

    fun select(id: String) {
        _state.value = _state.value.copy(selected = id)
    }
}
